### PR TITLE
Add device trust handler to teleterm daemon

### PR DESCRIPTION
**Pull request by avatus**
Based on https://github.com/gravitational/teleport/pull/40377

Contributes to https://github.com/gravitational/teleport.e/issues/3236

This PR adds the `AuthenticateWebDevice` handler to our daemon in order to exchange the `deviceWebToken` received during a deeplink being implemented shortly after https://github.com/gravitational/teleport/pull/40420 goes in

### CODE REVIEW

File: lib/teleterm/apiserver/handler/handler_device_trust.go [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+import (
+	"context"
+
+	api "github.com/gravitational/teleport/gen/proto/go/teleport/lib/teleterm/v1"
+	"github.com/gravitational/teleport/lib/teleterm/api/uri"
+	"github.com/gravitational/trace"
+)
+
+func (s *Handler) AuthenticateWebDevice(ctx context.Context, req *api.AuthenticateWebDeviceRequest) (*api.AuthenticateWebDeviceResponse, error) {
+	clusterURI, err := uri.Parse(req.GetRootClusterUri())
```
Comment (codingllama) [inline lib/teleterm/apiserver/handler/handler_device_trust.go:28]: Suggestion: run some preliminary validations here?

```go


	switch {
	case req.RootClusterUri == "":
		return nil, trace.BadParameter("root_cluster_uri required")
	case req.ConfirmationToken.GetId() == "":
		return nil, trace.BadParameter("confirmation_token.id required")
	case req.ConfirmationToken.GetToken() == "":
		return nil, trace.BadParameter("confirmation_token.token required")
	}
```

File: lib/teleterm/apiserver/handler/handler_device_trust.go [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	daemonResponse, err := s.DaemonService.AuthenticateWebDevice(ctx, clusterURI, req)
```
Comment (codingllama) [inline lib/teleterm/apiserver/handler/handler_device_trust.go:33]: Use just "resp" ?

```suggestion
	resp, err := s.DaemonService.AuthenticateWebDevice(ctx, clusterURI, req)
```

File: lib/teleterm/apiserver/handler/handler_device_trust.go [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return &api.AuthenticateWebDeviceResponse{ConfirmationToken: daemonResponse.ConfirmationToken}, nil
```
Comment (codingllama) [inline lib/teleterm/apiserver/handler/handler_device_trust.go:38]: Why the re-wrap? Could we do this:

```suggestion
	return resp, nil
```

File: lib/teleterm/daemon/daemon.go [hunk 4]
```diff
@@ -1074,6 +1076,25 @@ func (s *Service) UpdateUserPreferences(ctx context.Context, clusterURI uri.Reso
 	return preferences, trace.Wrap(err)
 }
 
+// AuthenticateWebDevice is used to upgrade a web session (identified by a DeviceWebToken) to include device trust extensions.
+func (s *Service) AuthenticateWebDevice(ctx context.Context, rootClusterURI uri.ResourceURI, req *api.AuthenticateWebDeviceRequest) (*api.AuthenticateWebDeviceResponse, error) {
+	proxyClient, err := s.GetCachedClient(ctx, rootClusterURI)
```
Comment (codingllama) [inline lib/teleterm/daemon/daemon.go:1081]: Personally I would write all of this in the base RPC handler (aka Handler. AuthenticateWebDevice) and do away with the changes here.

File: lib/teleterm/daemon/daemon.go [hunk 5]
```diff
@@ -32,11 +32,13 @@ import (
 
 	"github.com/gravitational/teleport/api/client/proto"
 	accesslistv1 "github.com/gravitational/teleport/api/gen/proto/go/teleport/accesslist/v1"
+	devicetrustv1 "github.com/gravitational/teleport/api/gen/proto/go/teleport/devicetrust/v1"
```
Comment (codingllama) [inline lib/teleterm/daemon/daemon.go:35]: Suggestion: I tend to use `devicepb`:

```suggestion
	devicepb "github.com/gravitational/teleport/api/gen/proto/go/teleport/devicetrust/v1"
```

File: lib/teleterm/daemon/daemon.go [hunk 6]
```diff
@@ -1074,6 +1076,25 @@ func (s *Service) UpdateUserPreferences(ctx context.Context, clusterURI uri.Reso
...
+// AuthenticateWebDevice is used to upgrade a web session (identified by a DeviceWebToken) to include device trust extensions.
+func (s *Service) AuthenticateWebDevice(ctx context.Context, rootClusterURI uri.ResourceURI, req *api.AuthenticateWebDeviceRequest) (*api.AuthenticateWebDeviceResponse, error) {
+	proxyClient, err := s.GetCachedClient(ctx, rootClusterURI)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	devicesClient := proxyClient.CurrentCluster().DevicesClient()
+
+	ceremony := dtauthn.NewCeremony()
+	confirmationToken, err := ceremony.RunWeb(ctx, devicesClient, &devicetrustv1.DeviceWebToken{Id: req.DeviceWebToken.Id, Token: req.DeviceWebToken.Token})
```
Comment (codingllama) [inline lib/teleterm/daemon/daemon.go:1088]: Suggestion: indent fields:

```suggestion
	confirmationToken, err := ceremony.RunWeb(ctx, devicesClient, &devicetrustv1.DeviceWebToken{
		Id: req.DeviceWebToken.Id,
		Token: req.DeviceWebToken.Token,
	})
```

**Review by codingllama — COMMENTED** [review]
Could we do a happy-path test? Doesn't have to be in-depth, for example we could fake the ceremony and just take a look at the input.

File: lib/teleterm/apiserver/handler/handler_device_trust.go (hunk 2 above)
Comment (avatus) [reply lib/teleterm/apiserver/handler/handler_device_trust.go]: Even better, I can just do 
```go
return s.DaemonService.AuthenticateWebDevice(ctx, clusterURI, req)
```

File: lib/teleterm/apiserver/handler/handler_device_trust.go [hunk 7]
```diff
@@ -25,15 +25,19 @@ import (
...
 	}
 
-	daemonResponse, err := s.DaemonService.AuthenticateWebDevice(ctx, clusterURI, req)
+	clusterURI, err := uri.Parse(req.GetRootClusterUri())
 	if err != nil {
 		return nil, trace.Wrap(err)
 	}
 
-	return &api.AuthenticateWebDeviceResponse{ConfirmationToken: daemonResponse.ConfirmationToken}, nil
+	return s.DaemonService.AuthenticateWebDevice(ctx, clusterURI, req)
```
Comment (codingllama) [inline lib/teleterm/apiserver/handler/handler_device_trust.go:42]: If you want to keep trace errors:

```suggestion
	resp, err := s.DaemonService.AuthenticateWebDevice(ctx, clusterURI, req)
	return resp, trace.Wrap(err)
```

**Review by codingllama — APPROVED** [review]

File: lib/teleterm/daemon/daemon.go (hunk 4 above)
Comment (ravicious) [reply lib/teleterm/daemon/daemon.go:1081]: We talked about this during the offsite. I had to add a new RPC and I was bemoaning our gRPC setup and Grzegorz said that you mentioned that as well in this PR.

I believe lib/teleterm was initially designed this way in order to decouple gRPC handlers from functions that perform the actual business logic (see https://github.com/gravitational/teleport/issues/14011#issuecomment-1195651304). At the time I had no experience working with gRPC, so I just stuck to the established convention.

Decoupling layers like that makes sense for HTTP routing, but nowadays I'm pretty sure it's unnecessary for gRPC. With HTTP, you have little control over what clients send to you, so it makes sense to have a layer dedicated to transforming an HTTP request into a valid message understood by your system. But with gRPC we already have protobufs, so this seems not needed. The messages sent with RCPs can work as inputs to your system. That's the "RPC" part of gRPC.

File: lib/teleterm/daemon/daemon.go [hunk 8]
```diff
@@ -1074,6 +1076,28 @@ func (s *Service) UpdateUserPreferences(ctx context.Context, clusterURI uri.Reso
...
 
+// AuthenticateWebDevice is used to upgrade a web session (identified by a DeviceWebToken) to include device trust extensions.
+func (s *Service) AuthenticateWebDevice(ctx context.Context, rootClusterURI uri.ResourceURI, req *api.AuthenticateWebDeviceRequest) (*api.AuthenticateWebDeviceResponse, error) {
+	proxyClient, err := s.GetCachedClient(ctx, rootClusterURI)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	devicesClient := proxyClient.CurrentCluster().DevicesClient()
+
+	ceremony := dtauthn.NewCeremony()
```
Comment (ravicious) [inline lib/teleterm/daemon/daemon.go:1087]: > Could we do a happy-path test? Doesn't have to be in-depth, for example we could fake the ceremony and just take a look at the input.

How do we fake the ceremony here though? Do we introduce a new interface implemented by two structs, one that does `dtauthn.NewCeremony()` and another which returns a fake ceremony in tests?

Even once that is solved, I'm not sure how we address `proxyClient.CurrentCluster().DevicesClient()`. Currently, some tests that don't actually need to execute stuff from clients are written to use `&client.ProxyClient{}` as `proxyClient`. I'm not sure if we'll be able to just call `CurrentCluster().DevicesClient()` on that.

https://github.com/gravitational/teleport/blob/2df6f6f8ccb3da16c64b9976828a410e2e480f83/lib/teleterm/daemon/daemon_test.go#L270-L278

https://github.com/gravitational/teleport/blob/2df6f6f8ccb3da16c64b9976828a410e2e480f83/lib/teleterm/daemon/daemon_test.go#L747-L749

**Review by ravicious — APPROVED** [review]

File: lib/teleterm/apiserver/handler/handler_device_trust.go [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+	"context"
+
+	api "github.com/gravitational/teleport/gen/proto/go/teleport/lib/teleterm/v1"
+	"github.com/gravitational/teleport/lib/teleterm/api/uri"
+	"github.com/gravitational/trace"
+)
+
+func (s *Handler) AuthenticateWebDevice(ctx context.Context, req *api.AuthenticateWebDeviceRequest) (*api.AuthenticateWebDeviceResponse, error) {
+	switch {
+	case req.GetRootClusterUri() == "":
```
[the remark this answers is not in the document]
Comment (codingllama) [reply lib/teleterm/apiserver/handler/handler_device_trust.go]: Just a note that the getters on req.DeviceWebToken (like `case req.DeviceWebToken.GetId() == ""`) are safe against a nil req.DeviceWebToken. If you drop those getters, make sure you req.DeviceWebToken before using it.

File: lib/teleterm/daemon/daemon.go (hunk 8 above)
Comment (codingllama) [reply lib/teleterm/daemon/daemon.go:1087]: Yes, that would take a bit more effort to make sure it all works. We could do `s.newDeviceAuthnCeremony()` (so we can swap that function) and we would need `proxyClient.CurrentCluster().DevicesClient()` to at least not panic.

The ceremony interface can be defined here, it only needs the RunWeb() method.

Alternatively we could use the fakes at [lib/devicetrust/testenv][1]. In that case we just need to make sure the DevicesClient() used is the one it provides. Test setup is pretty much the same as [TestCeremony_RunWeb][2].

No strong opinions on my part, whatever y'all think is best works for me here.

[1]: https://github.com/gravitational/teleport/blob/4713707b2d0b4541d5a6f21dc37c3b3128896dc9/lib/devicetrust/testenv/testenv.go#L69

[2]: https://github.com/gravitational/teleport/blob/4713707b2d0b4541d5a6f21dc37c3b3128896dc9/lib/devicetrust/authn/authn_test.go#L103

**Pull request merged**